examples/multi_agent/tic_tac_toe.py

```diff
@@ -0,0 +1,197 @@
+import os
+import pprint
+from torch.utils.tensorboard import SummaryWriter
+import torch
+import argparse
+import numpy as np
+from typing import Tuple, Optional
+
+from tianshou.env import MAEnv, VectorEnv
+from tianshou.policy import MADQNPolicy
+from tianshou.data import Collector, ReplayBuffer
+from tianshou.trainer import offpolicy_trainer
+
+from MANet import Net
+
+
+class TicTacToeEnv(MAEnv):
+    board = [[0, 0, 0],
+             [0, 0, 0],
+             [0, 0, 0]]
+
+    def __init__(self):
+        super().__init__(None)
+        self.current_board = np.array(TicTacToeEnv.board)
+        self.current_player = 0
+
+    def reset(self) -> dict:
+        self.current_board = np.array(TicTacToeEnv.board)
+        self.current_player = 0
+        return {
+            'agent_id': 0,
+            'obs': np.array(self.current_board),
+            'legal_actions': set(i for i in range(9))
+        }
+
+    def step(self, action: np.ndarray
+             ) -> Tuple[dict, np.ndarray, np.ndarray, dict]:
+        action = action.item(0)
+        assert 0 <= action < 9
+        assert self.current_board.item(action) == 0
+        self._move(action)
+        legal_actions = {i for i, b in enumerate(
```

Comment:
You could directly convert it to `list` since you are no doing any `set` specific operation later.

Reply:
This has to be ``set``, because available actions vary with time. If we hold it with list, they will be converted to ``np.ndarray`` and later be stacked, which will cause an exception (stacking arrays with inequal length).

Reply:
Which line are you updating `legal_actions` ?

Reply:
They are not updated. They becomes a field in observation.

Reply:
I don't tell you to use a `list` of course, but to use 
```
legal_actions = list({i for i, b in enumerate(
            self.current_board.flatten().tolist()) if b == 0})
```
since you are not relying on `set` properties elsewhere, and it would remove several conversion that comes later.

Reply:
> I don't tell you to use a `list` of course, but to use
> 
> ```
> legal_actions = list({i for i, b in enumerate(
>             self.current_board.flatten().tolist()) if b == 0})
> ```
> 
> since you are not relying on `set` properties elsewhere, and it would remove several conversion that comes later.

I tested your code, it raises an exception. You can try it, too :)

Reply:
I don't have time to check by myself, I trust you on that.

Reply:
> I don't have time to check by myself, I trust you on that.

You do not need to trust me. The code fails instantly, so you can try yourself and trust yourself :)

It seems strange: I think the conclusion is clear by checking the concept of Batch, but apparently it is not the case for you. Does it mean that we have some misunderstanding on the concept of ``Batch``?